pallets/evm-accounts/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+    dispatch::{Codec, DispatchInfo, Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    pallet_prelude::*,
+    traits::{tokens::Balance, Currency, IsSubType},
+};
+use frame_system::pallet_prelude::*;
+use pallet_transaction_payment::OnChargeTransaction;
+use sp_runtime::{
+    traits::{
+        Bounded, CheckedAdd, CheckedSub, Extrinsic, Hash, MaybeSerialize, Saturating,
+        SignedExtension, StaticLookup, Zero,
+    },
+    FixedPointNumber, FixedPointOperand,
+};
+use sp_std::{
+    convert::TryInto,
+    fmt::Debug,
+    marker::{Send, Sync},
+};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod test;
+
+mod eth;
+
+type BalanceOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use crate::eth::*;
+
+    use super::*;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The overarching call type.
+        type RuntimeCall: Parameter
+            + Dispatchable<
+                RuntimeOrigin = Self::RuntimeOrigin,
+                Info = DispatchInfo,
+                PostInfo = PostDispatchInfo,
+            > + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::RuntimeCall>;
+
+        /// The type of hash used for hashing the call.
+        type CallHasher: Hash;
+
+        /// The max number of substrate accounts that are linked to a given eth address.
+        type MaxLinkedAccounts: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub (super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Account have been linked to ethereum address
+        AccountLinked {
+            substrate: T::AccountId,
+            ethereum: EthAddress,
+        },
+        EthCallExecuted {
+            result: DispatchResult,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The ethereum address have an existing linkage already.
+        EthAddressAlreadyLinked,
+        /// The substrate address have an existing linkage already.
+        AccountAlreadyLinked,
+        /// The provided signature is invalid
+        BadSignature,
+        /// The substrate address is not linked to the given ethereum address
+        AccountNotLinked,
+        /// User have reached the maximum number of linked accounts.
+        CannotLinkMoreAccounts,
+    }
+
+    /// Map of one Substrate account to many EVM addresses
+    #[pallet::storage]
+    #[pallet::getter(fn evm_address_to_account)]
+    pub type Accounts<T: Config> = StorageMap<
```

Comment:
I don't like the storage name `Accounts`. It is not clear because EVM addresses are also EVM accounts. And then, without reading the doc for Accounts storage, it is not clear for which account this map is

Reply:
I renamed to `SubstrateAccounts`